### DISCUSSION TITLE
IMUSensor Data with specified frequency

### DISCUSSION THREAD

**Discussion by DanielAtt2000** [carla-simulator/carla#4163] [Q&A]
Hi,

I am trying to obtain IMU sensor data at a constant frequency of 50Hz i.e 50 data points for per second. After reading the docs and doing my best to understand them, I have reached to the conclusion that I should be using a fixed time-step for repeatability (still using the default Asynchronous mode) while varying the sensor_tick for the IMU sensor itself to achieve the 50Hz I require.

So I have set the sensor_tick to 1/50 = 0.02 and the fixed_delta_seconds to a constant 0.5.

I have created the IMU Sensor as was done in the manual_control.py example but unfortunately my understanding isn't sound since the sensor_tick does not effect the frequency of the data collection. Only when changing the fixed_delta_seconds to 0.02 does the IMU sensor be called 50 times a second. This results in a unrealistically slow simulation and high IMU sensor values (don't know the reason for this either). I have also tried enabling -no-rendering with no luck. Below I have included the IMU Class code.

    class CustomImuSensor(object):

    def __init__(self, parent_actor):
        self.sensor = None
        self._parent = parent_actor
        self.accelerometer = (0.0, 0.0, 0.0)
        self.gyroscope = (0.0, 0.0, 0.0)
        self.compass = 0.0
        world = self._parent.get_world()
        bp = world.get_blueprint_library().find('sensor.other.imu')
        bp.set_attribute("sensor_tick",str(0.02))
        print("IMU2" + str(bp.get_attribute("sensor_tick")))
        self.sensor = world.spawn_actor(bp, carla.Transform(), attach_to=self._parent)
        self.sensor.listen(lambda sensor_data: CustomImuSensor.custom_IMU_callback(sensor_data))

    @staticmethod
    def custom_IMU_callback(sensor_data):
        print("IMU2")
        array_mani2.add_imu_data(sensor_data,sensor_data.timestamp,0,sensor_data.frame)

Would anyone know where my understanding is flawed? I have a low-end machine but I'm not quite sure if this should have any effect.

Thank you for any help.
Daniel

**Comment by DSantosO** [comment]
Hello @DanielAtt2000,
As you have discover, the sensor_tick parameter can not be used to increase the frequency of the sensor higher than the actual ticking of the simulator. 
Anyway, decreasing the dt to 0.02 should work even if it is much slower and should give similar IMU values, can you please post the values you obtain in the two cases?

**Reply by DanielAtt2000** [reply]
> Hello @DanielAtt2000,
> As you have discover, the sensor_tick parameter can not be used to increase the frequency of the sensor higher than the actual ticking of the simulator.
> Anyway, decreasing the dt to 0.02 should work even if it is much slower and should give similar IMU values, can you please post the values you obtain in the two cases?

Hi @DSantosO,

Thank you for your reply.

I have exported the IMU sensor data to an excel file and created a line chart. Both charts exhibit the same type of manoeuvre, an acceleration followed by a braking event, braking at the same point and covering the same distance in total. The delta seconds is the only difference between these two simulations.

![image](https://user-images.githubusercontent.com/59541958/117148835-e99d8600-adb6-11eb-9f59-214e9665dc77.png)


![image](https://user-images.githubusercontent.com/59541958/117148328-6bd97a80-adb6-11eb-8754-1d35255354d1.png)


I can also say that the 0.02 delta seconds version runs much slow (obv) and therefore I keep pressing the acceleration arrow much more which may explain the higher acceleration values, but I'm not sure.

Does the FPS make any difference? For the dt = 0.5 server side fps = 9/10 while for dt = 0.02 server side fps = 13/14. Client side fps are relatively constant for both simulations at around 60/62 fps.

Also, after doing some research with other real life datasets the dt= 0.5 values are much closer to real life values where accelerations peak around at 4m/s^2.

Thank you for any help

**Reply by DSantosO** [reply]
Hello @DanielAtt2000,
Are you using keyboard input? 
Notice that using asynchronous mode and keyboard input you cannot compare both simulations even with fixed delta time. Be aware that in this mode you cannot control at which server frame exactly the controls are applied, only when the client sends them. This means that the server will receive and apply input after you stop sending them at the client. 
If you want to compare both simulations you should be very careful with the synchronization between client and server (using sync mode or using the `wait_for_tick()` method in async mode. Also, you need to be careful that you are applying the input the same interval of time.

**Reply by DanielAtt2000** [reply]
Thank you for the information. Yup that would be it!

**Comment by rahuldeo2047** [comment]
I tried with 500Hz IMU in sim_time with 9.15. It works but wonder why my RTX4090 is still under used and the simulation speed is around 3/27 (will be back if I remember incorrectly)
Is it same with the 0.10.0???
